macros/src/parser/lifetimes.rs

```diff
@@ -0,0 +1,98 @@
+use quote::quote;
+use std::ops::Deref;
+use syn::{parse, spanned::Spanned, GenericArgument, Lifetime, PathArguments, Type};
+
+#[derive(Debug, Clone)]
+pub struct Lifetimes(Vec<Lifetime>);
+
+impl Lifetimes {
+    pub fn new() -> Lifetimes {
+        Lifetimes(Vec::new())
+    }
+
+    pub fn from_type(data_type: &Type) -> Result<Lifetimes, parse::Error> {
+        let mut lifetimes = Lifetimes::new();
+        get_lifetimes(data_type, &mut lifetimes)?;
+        Ok(lifetimes)
+    }
+
+    pub fn push(&mut self, lifetime: &Lifetime) {
+        if !self.0.contains(lifetime) {
+            self.0.push(lifetime.to_owned());
+        }
+    }
+
+    pub fn append(&mut self, lifetimes: &Lifetimes) {
+        for lifetime in lifetimes.0.iter() {
+            self.push(lifetime);
+        }
+    }
+
+    pub fn remove(&mut self, lifetimes: &Lifetimes) {
+        self.0.retain(|lt| !lifetimes.0.contains(lt));
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn to_code(&self) -> proc_macro2::TokenStream {
+        if self.is_empty() {
+            quote! {}
+        } else {
+            let lifetimes = &self.0;
+            quote! { #(#lifetimes),* ,}
+        }
+    }
+}
+
+impl Default for Lifetimes {
+    fn default() -> Self {
+        Lifetimes::new()
+    }
+}
+
+impl Deref for Lifetimes {
+    type Target = [Lifetime];
+
+    fn deref(&self) -> &[Lifetime] {
+        &self.0[..]
+    }
+}
```

Comment:
Why do we need `Deref`? Seems somewhat opaque to me

Reply:
When `to_code` is under `lifetimes.rs`, we do not need it. But if I move it to `codegen.rs` we will need a way to get access to `&[Lifetime]`. I can remove it and add `to_slice` method if you prefer.